README.md
# letsencrypt-kubernetes

A docker image suitable for requesting new certificates from letsencrypt,
and storing them in a secret on kubernetes.

Available on docker hub as [ployst/letsencrypt](https://hub.docker.com/r/ployst/letsencrypt)

## Purpose

To provide an application that owns certificate requesting and storing.

- To serve acme requests to letsencrypt (given that you direct them to this
   container)
- To regularly (monthly) ask for new certificates.
- To store those new certificates in a secret on kubernetes.

## Useful commands

### Generate a new set of certs

Once this container is running you can generate new certificates using:

```bash
kubectl exec -it <pod> -- bash -c 'EMAIL=fred@fred.com DOMAINS=example.com foo.example.com ./fetch_certs.sh'
```

### Save the set of certificates as a secret

```bash
kubectl exec -it <pod> -- bash -c 'DOMAINS=example.com foo.example.com ./save_certs.sh'
```

### Refresh the certificates

```bash
kubectl exec -it <pod> -- bash -c 'EMAIL=fred@fred.com DOMAINS=example.com foo.example.com SECRET_NAME=foo DEPLOYMENTS=bar ./refresh_certs.sh'
```

## Environment variables:

- EMAIL
  - the email address to obtain certificates on behalf of.
- DOMAINS
  - a space separated list of domains to obtain a certificate for.
- LETSENCRYPT_ENDPOINT
  - If set, will be used to populate the /etc/letsencrypt/cli.ini file with
    the given server value. For testing use
    https://acme-staging.api.letsencrypt.org/directory
- DEPLOYMENTS
  - a space separated list of deployments whose pods should be refreshed after a certificate save
- SECRET_NAME
  - the name to save the secrets under
- NAMESPACE
  - the namespace under which the secrets should be available
- TYPE
  - the type of the secrets (default is Opaque)
- CRON_FREQUENCY
  - the 5-part frequency of the cron job. Default is a random time in the range `0-59 0-23 1-27 * *`
- SSL_VERIFY_URL
  - the update job will verify this service to be sure the certificate has been updated correctly. If the certificate has not been updated, and still will expire within ```${SSL_VERIFY_DAYS}``` days, a mail will be sent to ```${EMAIL_TO}``` recipients.
    This variable has the "hostname:port" format.
- SSL_VERIFY_DAYS
  - the number of days to check before the certificate will expire,  when the update job will check ```${SSL_VERIFY_URL}``` service.
    If not specified, the default value is 5 days.
- EMAIL_HOST
  - hostname of the mail server used to send the mail
- EMAIL_PORT
  - port of the mail server used to send the mail
- EMAIL_USERNAME
  - username used to login on then server
- EMAIL_PASSWORD
  - password used to login on then server
- EMAIL_FROM
  - email used to populate the FROM header of the message
- EMAIL_TO
  - a list of addresses to send the mail to, separated by space


CHANGELOG.md
# 0.5.0

 - Using Debian 10 Buster
 - Upgrade certbot to 0.31.0
 - Upgrade kubectl to latest (always)

# 0.4.0

 - Changed certificae update logic
 - Upgrade certbot to 0.21.1
 - Upgrade kubectl to latest (always)

# 0.3.0

 - Upgrade kubectl to 1.6.3

# 0.1.0

 - Upgrade kubectl to 1.3.6

# 0.0.9

 - NAMESPACE bugfixes

# 0.0.8

 - NAMESPACE env var support added

# 0.0.7

 - Upgrade kubectl so that deployments work

# 0.0.6

 - Added support for k8s deployments
 - Removed support for RC based refreshes

# 0.0.5


save_certs.sh
#!/bin/bash

# $DOMAINS should contain all domains that this container is responsible for
# renewing. The first one dictates where the cert will live.

# Inside /etc/letsencrypt/live/<domain> we have:
#
# cert.pem  chain.pem  fullchain.pem  privkey.pem
#
# We want to convert fullchain.pem into proxycert
# and privkey.pem into proxykey and then save as a secret!

if [ -z "${SECRET_NAME}" ]; then
    echo "ERROR: Secret name is empty or unset"
    exit 1
fi

CERT_LOCATION='/etc/letsencrypt/live'

DOMAINS=($DOMAINS)

DOMAIN=${DOMAINS[0]}

CERT=$(cat ${CERT_LOCATION}/${DOMAIN}/fullchain.pem | base64 --wrap=0)
KEY=$(cat ${CERT_LOCATION}/${DOMAIN}/privkey.pem | base64 --wrap=0)
PEM=$(cat ${CERT_LOCATION}/${DOMAIN}/fullchain.pem ${CERT_LOCATION}/${DOMAIN}/privkey.pem | base64 --wrap=0)
DHPARAM=$(openssl dhparam 2048 | base64 --wrap=0)

NAMESPACE=${NAMESPACE:-default}
TYPE=${TYPE:-kubernetes.io/tls}

kubectl get secrets --namespace ${NAMESPACE} ${SECRET_NAME} && ACTION=replace || ACTION=create;

cat << EOF | kubectl ${ACTION} -f -
{
  "apiVersion": "v1",
  "kind": "Secret",
  "metadata": {
    "annotations": {
      "kubernetes.io/tls-acme": "true"
    },
    "name": "${SECRET_NAME}",
    "namespace": "${NAMESPACE}"
  },
  "data": {
    "proxycert": "${CERT}",
    "proxykey": "${KEY}",
    "tls.crt": "${CERT}",
    "tls.key": "${KEY}",
    "tls.pem": "${PEM}",
    "dhparam": "${DHPARAM}"
  },
  "type": "${TYPE}"
}
EOF


fetch_certs.sh
#!/bin/bash

EMAIL=${EMAIL}
DOMAINS=(${DOMAINS})

if [ -z "${DOMAINS}" ]; then
    echo "ERROR: Domain list is empty or unset"
    exit 1
fi

if [ -z "${EMAIL}" ]; then
    echo "ERROR: Email is empty string or unset"
    exit 1
fi

CERT_LOCATION='/etc/letsencrypt/live'

DOMAIN=${DOMAINS[0]}

DAYS=15
TIME=$((60 * 60 * 24 * ${DAYS}))

MUST_UPGRADE="n"
if [ ! -f ${CERT_LOCATION}/${DOMAIN}/fullchain.pem ] || ! openssl x509 -noout -checkend ${TIME} -in ${CERT_LOCATION}/${DOMAIN}/fullchain.pem >/dev/null; then
    MUST_UPGRADE="y"
fi

if [ "${MUST_UPGRADE}" = "y" ]; then
    domain_args=""
    for i in "${DOMAINS[@]}"
    do
    domain_args="${domain_args} -d $i"
    # do whatever on $i
    done

    certbot certonly \
        --renew-by-default --agree-tos --email ${EMAIL} \
        --max-log-backups 0 \
        --webroot -w /letsencrypt/challenges/ \
        ${domain_args}

    echo "UPDATED" > /tmp/.letsencrypt-status
    exit 0
fi

echo "SKIPPED" > /tmp/.letsencrypt-status
exit 0


recreate_pods.sh
#!/bin/bash

# Update the required env vars for the first pod in each Deployment.
# This will kick off a rolling update.
# Do this so that the secrets can be remounted.

if [ -z "${DEPLOYMENTS}" ]; then
    echo "WARNING: DEPLOYMENTS not provided. Secret changes may not be reflected."
    exit
fi

DEPLOYMENTS=(${DEPLOYMENTS})
DATE=$(date)
NAMESPACE=${NAMESPACE:-default}

for DEPLOYMENT in "${DEPLOYMENTS[@]}"
do
    NAME=$(kubectl get deployments --namespace ${NAMESPACE} ${DEPLOYMENT} -o=template --template='{{index .spec.template.spec.containers 0 "name"}}')
    PATCH=$(NAME=${NAME} DATE=${DATE} echo "{\"spec\": {\"template\": {\"spec\": {\"containers\": [{\"name\": \"${NAME}\", \"env\": [{\"name\": \"LETSENCRYPT_CERT_REFRESH\", \"value\": \"${DATE}\"}]}]}}}}")
    echo "PATCHING ${DEPLOYMENT}: ${PATCH}"
    kubectl patch deployment --namespace ${NAMESPACE} ${DEPLOYMENT} --type=strategic -p "${PATCH}"
done


refresh_certs.sh
#!/bin/bash
set -e

if [ -e "/tmp/.letsencrypt-lock" ]; then
    OLDPID=$(cat /tmp/.letsencrypt-lock)
    if ps -p $OLDPID > /dev/null; then
        echo "Nope, not gonna touch that."
        exit 1
    fi
    rm -f /tmp/.letsencrypt-lock
fi

touch /tmp/.letsencrypt-lock
echo $BASHPID > /tmp/.letsencrypt-lock

echo "$(date) Fetching certs..."
/letsencrypt/fetch_certs.sh
if [ -f /tmp/.letsencrypt-status ]; then
  UPDATED=$(cat /tmp/.letsencrypt-status)
  rm /tmp/.letsencrypt-status
fi

if [ "${UPDATED}" = "UPDATED" ]; then
    echo "$(date) Saving certs..."
    /letsencrypt/save_certs.sh

    echo "$(date) Recreating pods..."
    /letsencrypt/recreate_pods.sh
else
    echo "$(date) The certificate han not been updated"
fi

rm /tmp/.letsencrypt-lock

if [ "${SSL_VERIFY_URL}" != "" ]; then
    NAMESPACE=${NAMESPACE:-default}
    SSL_VERIFY_DAYS=${SSL_VERIFY_DAYS:-5}
    TIME=$((60 * 60 * 24 * ${SSL_VERIFY_DAYS}))
    if ! echo "\n" | openssl s_client -connect ${SSL_VERIFY_URL} 2>/dev/null | \
                openssl x509 -noout -checkend ${TIME} > /dev/null 2>&1; then

        TO_PARAMS=""
        for address in "${EMAIL_TO[@]}"; do
            TO_PARAMS="${TO_PARAMS} --to ${address}"
        done

        gomailer ${TO_PARAMS} \
            --subject "[Cinello ${NAMESPACE} SSL] Expiring certificate on proxy server" \
            --body "WARNING! The SSL certificate on ${NAMESPACE} proxy server will expire in 5 days or less.\nPlease check it."
    fi
fi


start.sh
#!/bin/bash

# Add a cron line with details of the current user etc
minute=$(echo $RANDOM % 60 | bc)
hour=$(echo $RANDOM % 23 | bc)

CRON_FREQUENCY="$minute $hour * * *"
NAMESPACE=${NAMESPACE:-default}

echo "Configuring cron..."
echo "DOMAINS:        ${DOMAINS}"
echo "EMAIL:          ${EMAIL}"
echo "DEPLOYMENTS:    ${DEPLOYMENTS}"
echo "NAMESPACE:      ${NAMESPACE}"
echo "SECRET_NAME:    ${SECRET_NAME}"
echo "CRON frequency: ${CRON_FREQUENCY}"
# Once a month, fetch and save certs + restart pods.

# The process running under cron needs to know where the to find the kubernetes api
env_vars="PATH=${PATH} KUBERNETES_PORT=${KUBERNETES_PORT} KUBERNETES_PORT_443_TCP_PORT=${KUBERNETES_PORT_443_TCP_PORT} KUBERNETES_SERVICE_PORT=${KUBERNETES_SERVICE_PORT} KUBERNETES_SERVICE_HOST=${KUBERNETES_SERVICE_HOST} KUBERNETES_PORT_443_TCP_PROTO=${KUBERNETES_PORT_443_TCP_PROTO} KUBERNETES_PORT_443_TCP_ADDR=${KUBERNETES_PORT_443_TCP_ADDR} KUBERNETES_PORT_443_TCP=${KUBERNETES_PORT_443_TCP}"

line="${CRON_FREQUENCY} ${env_vars} SECRET_NAME=${SECRET_NAME} NAMESPACE=${NAMESPACE} DEPLOYMENTS='${DEPLOYMENTS}' DOMAINS='${DOMAINS}' EMAIL=${EMAIL} /bin/bash /letsencrypt/refresh_certs.sh >> /var/log/cron-encrypt.log 2>&1"
(crontab -u root -l; echo "${line}" ) | crontab -u root -

if [ -n "${LETSENCRYPT_ENDPOINT+1}" ]; then
    echo "server = ${LETSENCRYPT_ENDPOINT}" >> /etc/letsencrypt/cli.ini
fi

echo "Registering user..."
certbot register --email ${EMAIL} --no-eff-email --agree-tos --max-log-backups 0 || true

# Start cron
echo "Starting cron..."
cron &

echo "Starting nginx..."
nginx -g 'daemon off;'
